### DISCUSSION TITLE
How to downcast `dyn Array` using `as_any` and `data_type`?

### DISCUSSION THREAD

**Discussion by twitu** [jorgecarleitao/arrow2#1101] [Q&A] [answered]
I'm reading a chunk from a file, then iterating the column array but I need to convert the array to a more specific array type. I've looked at the documentation but haven't found a way to do it.

```rust
    let f = File::open("array_of_arrays.parquet").unwrap();
    let fr = FileReader::try_new(&f, None, None, None, None).unwrap();

    for chunk in fr.into_iter() {
        if let Ok(cols) = chunk {
            for array in cols.arrays().iter() {
                 // QUESTION: array: &Arc<dyn Array>
                 // QUESTION: cast array to some array type say ListArray
            }
        }
    }
```

Documentation for [data_type](https://docs.rs/arrow2/latest/arrow2/array/trait.Array.html#tymethod.data_type) mentions using `as_any` and `data_type` to this. But it's not very clear how to use this to downcast the array.

Any help is deeply appreciated.

**Comment by jorgecarleitao** [comment, marked as answer]
Hey! So, because `Array` is a trait object, we need to match the datatype as you mention. We tried to describe this in [the guide](https://jorgecarleitao.github.io/arrow2/high_level.html#downcast-and-as_any). Something like this:

```rust
use arrow2::datatypes::{PhysicalType, PrimitiveType};

match array.data_type().to_physical_type() {
     PhysicalType::Primitive(PrimitiveType::Float32) => {
            let array = array.as_any().downcast_ref::<PrimitiveArray<f32>>().unwrap(); // guaranteed
            // let array = f32-specific operator
            let array = array.clone();
            Ok(Box::new(array))
     }
     PhysicalType::Primitive(PrimitiveType::Float64) => {
            let array = array.as_any().downcast_ref::<PrimitiveArray<f64>>().unwrap(); // guaranteed
            // let array = f64-specific operator
            let array = array.clone();
            Ok(Box::new(array))
     }
     PhysicalType::List => {
            let array = array.as_any().downcast_ref::<ListArray<i32>>().unwrap(); // guaranteed
            // let array = list-specific operator
            let array = array.clone();
            Ok(Box::new(array))
     }
}
```

in other words, `data_type` is the logical type, `to_physical_type` converts it to its corresponding physical type, which we match to get the arrays' physical type.

Hope this helps.

**Reply by twitu** [reply]
Thanks for sharing this. I saw it in the documentation but it didn't click until now. Using this I've been able to successfully de-serialize parquet files back to structs.
